### PR TITLE
Screen space reflection

### PR DESCRIPTION
**Pull request by spiralhalo**
Adds screen space reflection shader that uses a view space normal color attachment, the depth buffer that comes with fabulous rendering, and projection matrix as inputs. Screen space reflection is computed based on roughness and metal fragment components. Despite the addition of those components, no serious PBR calculation is happening, just an approximation to get good enough visuals. The default shader has a few artifacts as a trade off for decent performance, but it can be replaced by third party shaders provided they follow the same workflow.

Summary of additions:

- Config for screen space reflection (disabled by default)
- Metal and roughness FragmentData components (accessible via shaders only in their current form)
- `material_main` draws metal and roughness component to extras buffer (replaces emissive buffer)
- `material_main` now draws to view space normal buffer
- CanvasFrameBufferHacks has an extra function to draw screen space reflection
- Edits to CanvasFrameBufferHacks to separate bloom and reflection processes
- Edits to `upsample` shader to allow blurring without addition
- FramebufferExt has an extra function to retrieve depth attachment
- `reflection_color` shader that computes the reflected color
- `reflection` shader that applies blurred reflected color to the base color
- Test materials: `glossy` for diamond and emerald blocks, `glossy_metal` for gold and iron blocks, and `rough_metal` for netherite block

p.s. The function that adds randomness to reflection rays (to simulate roughness) can be improved by using world space position as a seed, but it requires the inverted view matrix and I don't know how to obtain it.

### CODE REVIEW

File: src/main/resources/assets/canvas/shaders/internal/process/reflection.frag [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+varying vec2 _cvv_texcoord;
+
+void main() {
+    vec4 base 	    = texture2D(_cvu_base, _cvv_texcoord);
+	vec4 reflection = texture2D(_cvu_reflection, _cvv_texcoord);
+	float metal     = texture2DLod(_cvu_extras, _cvv_texcoord, 0).g;
+	float gloss     = 1 - texture2DLod(_cvu_extras, _cvv_texcoord, 0).b;
+    vec4 color      = mix(base, max(base, reflection), gloss * 0.6);
+    // vec4 metalColor1= mix(base, reflection, 0.5);
```
Comment (fewizz) [inline src/main/resources/assets/canvas/shaders/internal/process/reflection.frag:21]: wrong indentation?

File: src/main/resources/assets/canvas/shaders/internal/process/reflection_color.frag [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+	return 2.0 * texture2DLod(_cvu_normal, uv, 0).xyz - 1.0;
+}
+
+// Adds randomness to reflection ray for rough surfaces
+// Would look better if world space coordinates are used for seed
+const float wildness = 0.5;
+vec3 hash(vec2 uv){
+    vec3 a  = fract(uv.xyx * vec3(25.9, 25.9, 25.9));
+    a      += dot(a, a.yxz + 19.19);
+    vec3 b  = fract((a.xxy + a.yxx)*a.zyx);
```
Comment (fewizz) [inline src/main/resources/assets/canvas/shaders/internal/process/reflection_color.frag:41]: too

**Comment by grondag** [comment]
I'm excited to have a look at this. :-)

Will dig in after I finish some larger fixes in progress for item/GUI rendering.

**Comment by grondag** [comment]
One note I should have shared earlier: my plan for frame buffer attachments is currently as follows: 

color/alpha (existing main)
normal/height (new)
emissive/ao (extension of existing emissive attachment with full RGB. Possibly a float buffer to support HDR?)
metal/roughness/customA/customB (new, customA and customB channels available for 3rd party extensions/future additions)

Open to feedback or other ideas.  Not based on any testing.

**Comment by grondag** [comment]
Went through it very briefly online (not in IDE yet) and I like how it is put together and the direction it takes.  

There are some improvements I  already wanted to make for framebuffer/attachment management and I may try to slide some of those in with this.  And I may go ahead and set up the entire attachment structure so that it's available for contributors.

**Comment by spiralhalo** [comment]
Thank you for looking at my code. I'm glad that you're excited. And thanks @fewizz for marking the parts that needs reformatting as well.

I can understand about the framebuffer management. The framebuffer hacks and process shader classes need some refactoring before they turn into spaghetti code when more processes are added. I thought about doing it but I think the class structure is too big of a decision for me to make. I'm certainly happy if you would set up the attachment structure and/or the fragment data components along with it. Really looking forward to do more deferred rendering related stuff.

**Comment by spiralhalo** [comment]
The configurable pipeline pretty much made this PR obsolete with how flexible it is in terms of adding deferred shading features including one proposed in this PR. I'll close it for now and best of luck on the configurable pipeline development!

Still looking forward to those PBR params tho :)

**Pull request closed without merging**